### PR TITLE
Propery extensions

### PR DESCRIPTION
**Pull request by JensWehner**
Made the string to type conversion much more powerful. Now `as<type>()` can be `bool` any numeric type or any kind of object that is constructible from a string and also a `std::vector<T>` of any of the above mentioned types, also Eigen::VectorXd Eigen::VectorXf and Eigen::Vector3d and Eigen::Vector3f are supported. 

Furthermore you can now delete Attributes and Children, although the later invalidates all pointers to child nodes. 

If we wanted we could go back to use a std::list underneath, instead of a std::vector and then all references would be valid, but then large xml files would take ages to load.

### CODE REVIEW

**Review by junghans — APPROVED** [review]

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@votca-bot changelog: properties can be removed and string conversion much improved

File: src/libtools/property.cc [hunk 1]
```diff
@@ -50,15 +52,15 @@ const Property &Property::get(const string &key) const {
   }
 
   const Property *p;
-  map<string, Index>::const_iterator iter;
+  map<string, std::vector<Index>>::const_iterator iter;
```
Comment (felipeZ) [inline src/libtools/property.cc:55]: There is now a vector of indices in order to represent multiple attributes, right?

File: src/libtools/property.cc [hunk 2]
```diff
@@ -73,6 +75,10 @@ const Property &Property::get(const string &key) const {
   return *p;
 }
 
+Property &Property::get(const string &key) {
+  return const_cast<Property &>(static_cast<const Property &>(*this).get(key));
```
Comment (felipeZ) [inline src/libtools/property.cc:79]: Could you please briefly explain to me why is this double-casting needed? :)

File: src/libtools/property.cc [hunk 3]
```diff
@@ -185,6 +187,37 @@ std::vector<Property *> Property::Select(const string &filter) {
...
+  if (indices.size() == 1) {
+    _map.erase(prop.name());
+  } else {
+    indices.erase(std::remove(indices.begin(), indices.end(), index_of_child),
+                  indices.end());
+  }
+
+  // if child is not the last element we have to do two things, a) switch the
+  // child with the last element and b) update the index of the former last
+  // element
```
Comment (felipeZ) [inline src/libtools/property.cc:208]: Awesome comment here ;)

**Review by felipeZ — APPROVED** [review]

File: src/libtools/property.cc (hunk 1 above)
Comment (JensWehner) [reply src/libtools/property.cc:55]: yes to have indeces to all the properties which have the same tag.

File: src/libtools/property.cc (hunk 2 above)
Comment (JensWehner) [reply src/libtools/property.cc:79]: Basically it removes all the `const` , I think it was inspired by stack overflow.

**Comment by JensWehner** [comment]
@votca-bot format

**Review by felipeZ — APPROVED** [review]

**Pull request merged**